src/Extension.Tests/CodeRunEvaluationTests.cs

```diff
@@ -0,0 +1,290 @@
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            await kernel.SubmitCodeAsync(@"
+#!question 5
+
+double triangleArea(double x, double y) 
+{
+    return 0.5 * x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(true);
+        }
+
+
+        [Fact]
+        public async Task evaluation_criteria_is_matched_to_the_correct_question()
+        {
+            var evaluator = new Evaluator();
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+            await kernel.SubmitCodeAsync(@"
+#!answer 1
+
+area(10, 10) == 100
+");
+
+            await kernel.SubmitCodeAsync(@"
+#!answer 2
+
+area(10, 10) == 50
+");
+
+            var resultQuestion2 = await kernel.SubmitCodeAsync(@"
+#!question 2
+
+double area(double x, double y) 
+{
+    return x * y;
+}
+");
+
+            var resultQuestion1 = await kernel.SubmitCodeAsync(@"
+#!question 1
+
+double area(double x, double y) 
+{
+    return 0.5 * x * y;
+}
+");
+
+            resultQuestion1.KernelEvents.ToSubscribedList()
+                .Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(false);
+
+            resultQuestion2.KernelEvents.ToSubscribedList()
+                .Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(false);
+        }
+
+
+        [Fact]
+        public async Task multiple_evaluation_criteria_for_one_question_is_evaluated_as_a_conjunction()
+        {
+            var evaluator = new Evaluator();
+            using var csharpKernel = new CSharpKernel();
+            using var kernel = new CompositeKernel { csharpKernel };
+            kernel
+                .UseQuestionMagicCommand(evaluator)
+                .UseAnswerMagicCommand(evaluator);
+
+            using var events = kernel.KernelEvents.ToSubscribedList();
+
+            await kernel.SubmitCodeAsync(@"
+#!answer 1
+
+area(10, 10) == 50
+");
+
+            await kernel.SubmitCodeAsync(@"
+#!answer 1
+
+area(0, 0) == 0
+");
+
+            await kernel.SubmitCodeAsync(@"
+#!question 1
+
+double area(double x, double y) 
+{
+    return x * y;
+}
+");
+
+            events.Should().ContainSingle<CodeRunEvaluationProduced>()
+                .Which.Evaluation.Passed.Should().Be(false);
+        }
+
+
+        [Fact]
+        public async Task evaluation_criterion_that_does_not_have_ReturnValueProduced_does_not_affect_evaluation()
```

Comment:
It's probably more useful to describe what it _does_ affect. 

It's not clear to me from the test name what the intent is.

Reply:
Renamed to only_evaluation_criterion_that_returns_a_boolean_can_affect_evaluation.
The test illustrates the current behavior. The current behavior is that only criteria that return booleans are considered. Those that do not have ReturnValueProduced or whose return value is not bool are discarded before evaluation.